File: src/main/kotlin/xyz/xenondevs/nova/serialization/gson/EnumMapInstanceCreator.java
package xyz.xenondevs.nova.serialization.gson;

import com.google.gson.InstanceCreator;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.EnumMap;

// doesn't seem to work in kotlin
@SuppressWarnings({"unchecked", "rawtypes"})
public class EnumMapInstanceCreator implements InstanceCreator<EnumMap<?, ?>> {
    
    // https://stackoverflow.com/questions/54966118/how-to-deserialize-an-enummap
    public EnumMap<?, ?> createInstance(Type type) {
        return new EnumMap((Class<?>) ((ParameterizedType) type).getActualTypeArguments()[0]);
    }
    
}

File: src/main/kotlin/xyz/xenondevs/nova/serialization/gson/EnumMapInstanceCreator.kt
package xyz.xenondevs.nova.serialization.gson

import com.google.gson.InstanceCreator
import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type
import java.util.*

private val ENUMMAP_CONSTRUCTOR = EnumMap::class.java.getConstructor(Class::class.java)

class EnumMapInstanceCreator : InstanceCreator<EnumMap<*, *>> {
    
    // https://stackoverflow.com/questions/54966118/how-to-deserialize-an-enummap
    override fun createInstance(type: Type): EnumMap<*, *> {
        return ENUMMAP_CONSTRUCTOR.newInstance((type as ParameterizedType).actualTypeArguments[0] as Class<*>)
    }
    
}